packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -363,6 +394,63 @@ contract OPContractsManager is ISemver {
         return output;
     }
 
+    /// @notice Upgrades a set of chains to the latest implementation contracts
+    /// @param _opChains Array of OpChain structs, one per chain to upgrade
+    /// @dev This function is intended to be called via DELEGATECALL from the Upgrade Controller Safe
+    function upgrade(OpChain[] memory _opChains) external {
+        if (address(this) == address(thisOPCM)) revert OnlyDelegatecall();
+
+        Implementations memory impls = thisOPCM.implementations();
+        Blueprints memory prints = thisOPCM.blueprints();
+
+        // TODO: upgrading the SuperchainConfig and ProtocolVersions
+        // TODO: Check that all chains have the same SuperchainConfig
+
+        for (uint256 i = 0; i < _opChains.length; i++) {
+            ISystemConfig systemConfig = _opChains[i].systemConfigProxy;
+            ISystemConfig.Addresses memory opChainAddrs = _getAddresses(systemConfig);
+
+            IProxyAdmin proxyAdmin = _opChains[i].proxyAdmin;
+
+            // -------- Upgrade Contracts Stored in SystemConfig --------
+
+            // forgefmt: disable-start
+            proxyAdmin.upgradeAndCall(payable(address(systemConfig)), impls.systemConfigImpl, abi.encodeCall(ISystemConfig.upgrade, ()));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.l1CrossDomainMessenger), impls.l1CrossDomainMessengerImpl, abi.encodeCall(IL1CrossDomainMessenger.upgrade, (systemConfig)));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.l1StandardBridge), impls.l1StandardBridgeImpl, abi.encodeCall(IL1StandardBridge.upgrade, (systemConfig)));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.l1ERC721Bridge), impls.l1ERC721BridgeImpl, abi.encodeCall(IL1ERC721Bridge.upgrade, ()));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.disputeGameFactory), impls.disputeGameFactoryImpl, abi.encodeCall(IDisputeGameFactory.upgrade, ()));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.optimismPortal), impls.optimismPortalImpl, abi.encodeCall(IOptimismPortal2.upgrade, ()));
+            proxyAdmin.upgradeAndCall(payable(opChainAddrs.optimismMintableERC20Factory), impls.optimismMintableERC20FactoryImpl, abi.encodeCall(IOptimismMintableERC20Factory.upgrade, ()));
+
+            // -------- Discover and Upgrade Proofs Contracts --------
+            OpChainDiscoverables memory discoverables =
+                _getDiscoverables(IDisputeGameFactory(opChainAddrs.disputeGameFactory));
+
+            console.logBytes(blueprint.anchorStateRegistry.code);
+
+            address newAnchorStateRegistryImpl = Blueprint.deployFrom(
+                prints.anchorStateRegistry,
+                computeSalt(discoverables.l2ChainId, "salt mixer", "AnchorStateRegistry"),
+                abi.encode(opChainAddrs.disputeGameFactory)
+            );
```

Comment:
We should benchmark the cost of this method given the number of chains we'll want to upgrade at once for cycle 34, since deploying new contracts is expensive. Want to make sure that if we're near the block gas limit we know sooner rather than later